File: graph/google_Encoded_Dict.java
/**
 * google_Encoded_Dict.
 * google
 *
 * All words in a dictionary are decoded by mapping one character to another.
 * It's strictly one-to-one mapping, which means no multiple characters are mapped
 * to the same char.
 * For example:
 *
 * cat
 * coffee
 * common
 *
 * After the encoding:
 *
 * dkc
 * dbhhzz
 * dbllbq
 *
 * And of course this dictionary is in lexicographical order.
 *
 * Try to decode the dictionary.
 */


import java.util.HashSet;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.Stack;


public class google_Encoded_Dict {

  private static final int charNum = 26;

  public static void main(String[] args) {
    google_Encoded_Dict solution = new google_Encoded_Dict();
    solution.test();
  }

  public void test() {
    // Test case
    String[] dict = new String[] {"dkc", "dbhhzz", "dbllbq"};
    System.out.println("hello world");
    decode(dict);
  }

  public void decode(String[] dict) {
    int n = dict.length, index = 0, i;
    boolean printAll = true;
    char prev, cur;
    HashSet<String> relation = new HashSet<String>();
    while (printAll) {
      printAll = false;
      prev = '\0';
      for (i = 0; i < n; ++i) {
        if (index < dict[i].length()) {
          cur = dict[i].charAt(index);
          if (prev == '\0') {
            prev = cur;
          } else if (prev != cur) {
            // Build graph node
            relation.add(String.format("%s:%s", prev, cur));
            prev = cur;
          } else {
            // Keep looping.
            break;
          }
        }
        if (!printAll && dict[i].length() - 1 > index) {
          printAll = true;
        }
      }
      ++index;
    }
    int[][] graph = buildGraph(relation);
    topoSort(graph);
  }

  // Given the order relation, build graph.
  private int[][] buildGraph(Set<String> relation) {
    int[][] graph = new int[charNum][charNum];
    for (String str : relation) {
      String[] nodes = str.split(":");
      // a - 0, z - 25
      int v = nodes[0].charAt(0) - 'a';
      int t = nodes[1].charAt(0) - 'a';
      //System.out.println(String.format("%s %s", nodes[0], nodes[1]));
      graph[v][t] = 1;
    }

    System.out.println("hello world");
    printMatrix(graph);
    System.out.println("hello world");
    return graph;
  }

  // Topological Sorting.
  private void topoSort(int[][] graph) {
    Stack<Integer> s = new Stack<Integer>();
    boolean[] visited = new boolean[charNum];
    for (int i = charNum - 1; i >= 0; --i) {
      if (!visited[i]) {
        topoSortUtil(graph, i, visited, s);
      }
    }
    // stdout order.
    int i = 0;
    Map<Character, Character> encodingKey = new HashMap<Character, Character>();

    assert(s.size() == 26);

    while (!s.isEmpty()) {
      int num = s.pop();
      char encodedChar = (char) (num + 97);
      char decodedChar = (char) (i++ + 97);
      encodingKey.put(decodedChar, encodedChar);
    }

    System.out.println(encodingKey);
  }

  private static String[] decodeWithKey(String[] dict, HashMap<Character, Character> encodingKey) {
    //Given encoding mapping, decode the dicionary.
    // Decoding...
    return dict;
  }

  private void topoSortUtil(int[][] graph, int i, boolean[] visited, Stack<Integer> s) {
    visited[i] = true;

    for (int v = 0; v < charNum; ++v) {
      if (!visited[v] && graph[i][v] == 1) {
        topoSortUtil(graph, v, visited, s);
      }
    }
    s.push(i);
  }

  // Helper function to print.
  public void print(int[] arr) {
    for (int i : arr) System.out.print(i + " ");
    System.out.println();
  }

  public void printMatrix(int[][] matrix) {
    for (int[] arr : matrix) print(arr);
    System.out.println();
  }
}

File: graph/google_Encoded_Dict.py
"""google_Encoded_Dict.py
google

All words in a dictionary are decoded by mapping one character to another.
It's strictly one-to-one mapping, which means no multiple characters are mapped
to the same char.
For example:

cat
coffee
common

After the encoding:

dkc
dbhhzz
dbllbq

And of course this dictionary is in lexicographical order.
Try to decode the dictionary.

This is the Python version. Java version is: graph/google_Encoded_Dict.java
"""


class Solution(object):

    """"""

    def __init__(self, words):
        """"""
        self._words = words

    def solve(self):
        """"""
        graph = self.create_graph(self._words)

        mapping = graph.toposort()

        encode = lambda x: mapping[x]
        for word in self._words:
            print map(encode, word)

    def create_graph(self, words):
        """"""
        graph = Graph()

        # Vertical order. Think about alphabetical order. That's where we find
        # directed node connectivity.
        for i in xrange(len(words) - 1):
            cur_word = words[i]
            nex_word = words[i + 1]

            for j in xrange(max(len(cur_word), len(nex_word))):
                if cur_word[j] == nex_word[j]:
                    continue
                else:
                    print nex_word[j], cur_word[j]
                    graph.add_edge(nex_word[j], cur_word[j])
                    break

        return graph


class Graph(object):

    LENGTH=26

    """Graph class for lowercase alphabeta.

    Dimension is fixed at 26. Using adjacent matrix to represent the grapy.
    """

    def __init__(self):
        """"""
        self._matrix = [[0 for i in xrange(Graph.LENGTH)] for _ in xrange(Graph.LENGTH)]
        self._mapping = {chr(i + 97) : i for i in xrange(Graph.LENGTH)}
        self._indegree = [0] * Graph.LENGTH

    def add_edge(self, a, b):
        """"""
        if a == b:
            return

        i = self._mapping[a]
        j = self._mapping[b]

        print a, i, b, j
        if self._matrix[i][j] == 0:
            self._matrix[i][j] = 1
            self._indegree[j] += 1

    def toposort(self):
        """Iterative way to find the topo sort of a graph."""
        topo_order = []

        visited = [0] * Graph.LENGTH

        for i in xrange(Graph.LENGTH):
            v = -1
            # Find the current node with zero indegree.
            for j in xrange(Graph.LENGTH):
                if v >= 0:
                    break
                if self._indegree[j] == 0 and not visited[j]:
                    v = j
            if v < 0:
                break

            visited[v] = 1
            topo_order.append(v)

            for j in xrange(Graph.LENGTH):
                if self._matrix[v][j]:
                    self._indegree[j] -= 1

        return self.invert_mapping(topo_order)

    def invert_mapping(self, order):
        """"""
        assert len(set(order)) == len(order), 'Duplicates in topo order: {0}'.format(order)
        mapping = {}

        print order
        for i, val in enumerate(order):
            #print 'topo order'
            #print i, val
            #print chr(i + 97), chr(val + 97)
            mapping[chr(i + 97)] = chr(val + 97)

        return mapping

    def __repr__(self):
        """"""
        return 'Graph object at {0}'.format(hex(id(self)))


def main():
    words = ['dkc',
             'dbhhz',
             'dbllbq']

    Solution(words).solve()



if __name__ == '__main__':
    main()
